devices/src/virtio/mmio.rs

```diff
@@ -41,12 +41,47 @@ pub trait VirtioDevice: Send {
 
     /// The set of feature bits shifted by `page * 32`.
     fn features(&self, page: u32) -> u32 {
-        let _ = page;
-        0
+        let avail_features = self.get_available_features();
+        match page {
+            // Get the lower 32-bits of the features bitfield.
+            0 => avail_features as u32,
+            // Get the upper 32-bits of the features bitfield.
+            1 => (avail_features >> 32) as u32,
+            _ => {
+                warn!("Received request for unknown features page.");
+                0u32
+            }
+        }
     }
 
     /// Acknowledges that this set of features should be enabled.
-    fn ack_features(&mut self, page: u32, value: u32);
+    fn ack_features(&mut self, page: u32, value: u32) {
+        let mut v = match page {
+            0 => u64::from(value),
+            1 => u64::from(value) << 32,
+            _ => {
+                warn!("Cannot acknowledge unknown features page: {}", page);
+                0u64
+            }
+        };
+
+        // Check if the guest is ACK'ing a feature that we didn't claim to have.
+        let avail_features = self.get_available_features();
+        let unrequested_features = v & !avail_features;
+        if unrequested_features != 0 {
+            warn!("Received acknowledge request for unknown feature: {:x}", v);
+            // Don't count these features as acked.
+            v &= !unrequested_features;
+        }
+
+        let res = avail_features | v;
+        self.set_acked_features(res);
```

Comment:
Less code suggestion:
```suggestion
        self.set_acked_features(avail_features | v);
```